src/chain/chain.rs

```diff
@@ -0,0 +1,331 @@
+use std::borrow::Borrow;
+use std::marker::PhantomData;
+
+use hal::pso::PipelineStage;
+
+use chain::pass::PassLinks;
+use chain::link::{Acquire, Link, LinkSync, Release};
+use resource::{Access, Layout, Resource, Usage};
+
+/// Unique identifier for resource dependency chain.
+/// Multiple resource can be associated with single chain
+/// if all passes uses them the same way.
+/// Chain id uses marker type so that ids for buffers and images are different.
+#[derive(Derivative)]
+#[derivative(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+pub struct ChainId<T: ?Sized>(usize, PhantomData<T>);
+impl<T> ChainId<T> {
+    /// Make new chain id.
+    pub fn new(index: usize) -> Self {
+        ChainId(index, PhantomData)
+    }
+
+    /// Get index value.
+    pub fn index(&self) -> usize {
+        self.0
+    }
+}
+
+/// Resource chain for set of resources.
+/// All resources in set are expected to be used only in ways specified by this chain.
+/// Each link of the chain corresponds to some render-pass or similar entity.
+#[derive(Clone, Debug)]
+pub struct Chain<A, L, U, S, W> {
+    usage: U,
+    links: Vec<Option<Link<A, L, S, W>>>,
+}
+
+impl<A, L, U, S, W> Chain<A, L, U, S, W>
```

Comment:
we could parametrize the implementation by `R` right away, since the only method it has (`build`) defines `R`, and we can derive `A/L/U` with their bounds from it

Reply:
Compiller wouldprobably complain that `R` is not used in type signature.